engine/src/main/java/org/terasology/world/chunks/blockdata/ExtraBlockDataManager.java

```diff
@@ -0,0 +1,305 @@
+                        logger.error("Unable to register extra block data: "+errorType+" for {}.{}: should be \"public static boolean {}(Block block)\", and bitSize should be 4, 8 or 16.", type.getName(), method.getName(), method.getName());
+                        continue;
+                    }
+                    method.setAccessible(true);
+                    HashSet<Block> includedBlocks = new HashSet<>();
+                    for (Block block : blocks) {
+                        try {
+                            if ((boolean) method.invoke(null, block)) {
+                                includedBlocks.add(block);
+                            }
+                        } catch(IllegalAccessException e) {
+                            // This should not get to this point.
+                            throw new RuntimeException("Incorrect access modifier on register extra data method", e);
+                        } catch(InvocationTargetException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                    fieldss.get(registerAnnotation.bitSize()).put(registerAnnotation.name(), includedBlocks);
+                }
+            }
+        }
+        
+        slots = new HashMap<>();
+        ArrayList<TeraArray.Factory<?>> tempSlotTypes = new ArrayList<>();
+        fieldss.forEach((size, fields) -> {
+            Graph disjointnessGraph = getDisjointnessGraph(fields);
+            ArrayList<ArrayList<String>> cliques = findCliqueCover(disjointnessGraph);
+            for (ArrayList<String> clique : cliques) {
+                for (String label : clique) {
+                    slots.put(label, tempSlotTypes.size());
+                }
+                tempSlotTypes.add(teraArrayFactories.get(size));
+            }
+        });
+        slotFactories = tempSlotTypes.toArray(new TeraArray.Factory<?>[0]);
+        String loggingOutput = "Extra data slots registered:";
+        boolean first = true;
+        for (Map.Entry<String, Integer> entry : slots.entrySet()) {
+            loggingOutput += (first ? " " : ", ") + entry.getKey() + " -> " + entry.getValue();
+            first = false;
+        }
+        logger.info(loggingOutput);
+    }
+    
+    private static String validRegistrationMethod(Method method, RegisterExtraData annotation) {
+        Class<?>[] argumentTypes = method.getParameterTypes();
+        return method.getReturnType() != boolean.class ? "incorrect return type" :
+               //! method.isAccessible() ? "method not accessible" :
+               ! teraArrayFactories.containsKey(annotation.bitSize()) ? "invalid bitSize" :
+               ! Modifier.isStatic(method.getModifiers()) ? "method not static" :
+               argumentTypes.length != 1 ? "arguments list has wrong length" :
+               argumentTypes[0] != Block.class ? "incorrect argument type" :
+               null;
+    }
+    
+    private static class Graph {
```

Comment:
I'll get onto the javadoc shortly. This wasn't based on any existing code. I just thought that as this was (in this case) only used internally by ExtraBlockDataManager, and I only needed a few methods, it would just be easier to implement it there than find an existing implementation.